File: mongo-mcp-server/src/handlers/indexHandlers.js
import { defaultDatabaseName } from "../config.js";
import { getDatabase } from "../mongo.js";
import { textContent } from "../response.js";
import {
  validateCollectionName,
  validateIndexKeys,
  validateIndexName,
  validateIndexOptions
} from "../validators.js";

export const indexHandlers = {
  get_indexes: async ({ database = defaultDatabaseName, collection }) => {
    validateCollectionName(collection);

    const indexes = await getDatabase(database)
      .collection(collection)
      .indexes();

    return textContent(indexes);
  },

  create_index: async ({
    database = defaultDatabaseName,
    collection,
    keys,
    options = {}
  }) => {
    validateCollectionName(collection);
    validateIndexKeys(keys);
    validateIndexOptions(options);

    const indexName = await getDatabase(database)
      .collection(collection)
      .createIndex(keys, options);

    return textContent({
      indexName,
      database,
      collection,
      keys,
      options
    });
  },

  update_index: async ({
    database = defaultDatabaseName,
    collection,
    indexName,
    keys,
    options = {}
  }) => {
    validateCollectionName(collection);
    validateIndexName(indexName);
    validateIndexKeys(keys);
    validateIndexOptions(options);

    const mongoCollection = getDatabase(database).collection(collection);

    await mongoCollection.dropIndex(indexName);
    const newIndexName = await mongoCollection.createIndex(keys, options);

    return textContent({
      droppedIndexName: indexName,
      newIndexName,
      database,
      collection,
      keys,
      options
    });
  },

  delete_index: async ({
    database = defaultDatabaseName,
    collection,
    indexName
  }) => {
    validateCollectionName(collection);
    validateIndexName(indexName);

    const result = await getDatabase(database)
      .collection(collection)
      .dropIndex(indexName);

    return textContent({
      deletedIndexName: indexName,
      database,
      collection,
      result
    });
  }
};

File: python-mcp-workspace/mongo_mcp_server/handlers/index_handlers.py
from ..config import DEFAULT_DATABASE_NAME
from ..mongo import get_database
from ..serializers import to_jsonable
from ..validators import (
    validate_collection_name,
    validate_index_keys,
    validate_index_name,
    validate_index_options,
)


def get_indexes(
    collection: str,
    database: str = DEFAULT_DATABASE_NAME,
) -> list:
    validate_collection_name(collection)

    indexes = list(get_database(database)[collection].list_indexes())

    return to_jsonable(indexes)


def create_index(
    collection: str,
    keys: dict,
    database: str = DEFAULT_DATABASE_NAME,
    options: dict | None = None,
) -> dict:
    options = options or {}

    validate_collection_name(collection)
    validate_index_keys(keys)
    validate_index_options(options)

    index_name = get_database(database)[collection].create_index(
        list(keys.items()),
        **options,
    )

    return {
        "indexName": index_name,
        "database": database,
        "collection": collection,
        "keys": keys,
        "options": options,
    }


def update_index(
    collection: str,
    indexName: str,
    keys: dict,
    database: str = DEFAULT_DATABASE_NAME,
    options: dict | None = None,
) -> dict:
    options = options or {}

    validate_collection_name(collection)
    validate_index_name(indexName)
    validate_index_keys(keys)
    validate_index_options(options)

    collection_instance = get_database(database)[collection]
    collection_instance.drop_index(indexName)
    new_index_name = collection_instance.create_index(list(keys.items()), **options)

    return {
        "droppedIndexName": indexName,
        "newIndexName": new_index_name,
        "database": database,
        "collection": collection,
        "keys": keys,
        "options": options,
    }


def delete_index(
    collection: str,
    indexName: str,
    database: str = DEFAULT_DATABASE_NAME,
) -> dict:
    validate_collection_name(collection)
    validate_index_name(indexName)

    get_database(database)[collection].drop_index(indexName)

    return {
        "deletedIndexName": indexName,
        "database": database,
        "collection": collection,
        "result": True,
    }
